### PR TITLE
Fix pod deletion after a failed start under docker.

### PR DESCRIPTION
**Pull request by jcvenegas**
cc-runtime was failing to start a container with a non exiting container ( expected ) but delete cleanup was not properly handled.
This PR add fixes for:
https://github.com/clearcontainers/runtime/issues/507

1. Because the start the container failed, we should stop the shim (indicate the process is not running)
2. The runtime should use oci config.json only at create after that it uses its state file to perform following operations. We were saving the config.json path in virtcontainers annotations, lets save the json content. 
     This allow cc-runtime to get information related with the config even if does not exist anymore.

### CODE REVIEW

**Comment by jcvenegas** [comment]
I tested 
```
docker run --runtime cc-runtime -ti busybox asdf
```
no more shims or VMs are left but looking at runtime, `delete` still fails saying
```
cc-runtime:error:Could not get veth interface: LinkByName() failed for veth name eth0: Link not found
```
I debugged and it comes from 
https://github.com/containers/virtcontainers/blob/master/network.go#L322
https://github.com/containers/virtcontainers/blob/master/cnm.go#L197
https://github.com/containers/virtcontainers/blob/master/api.go#L131
 is there something I am missing related with network ?

**Comment by jcvenegas** [comment]
@sameo @devimc  could about why the vc does not find the veth pair on delete ?

**Comment by devimc** [comment]
@jcvenegas might be because it was not created? did you check that?

**Comment by jcvenegas** [comment]
@devimc  `create` is being called correctly, I added a `return err` at the end of a successful `start` and it failed as well my guess is docker is cleaning those things as well as happened with with config.json

Here the trace showhing `bridgeNetworkPair` was called correctly 
```
cc-runtime:info:Calling bridgework
[{"Index":1,"MTU":65536,"TxQLen":1000,"Name":"lo","HardwareAddr":null,"Flags":5,"RawFlags":65609,"ParentIndex":0,"MasterIndex":0,"Namespace":null,"Alias":"","Statistics":{"RxPackets":0,"TxPackets":0,"RxBytes":0,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"loopback","Protinfo":null,"OperState":0},{"Index":2,"MTU":1500,"TxQLen":1000,"Name":"tap0","HardwareAddr":"VppqtMYf","Flags":19,"RawFlags":4163,"ParentIndex":0,"MasterIndex":3,"Namespace":null,"Alias":"","Statistics":{"RxPackets":0,"TxPackets":0,"RxBytes":0,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"ether","Protinfo":null,"OperState":0,"LinkType":"tun"},{"Index":3,"MTU":1500,"TxQLen":0,"Name":"br0","HardwareAddr":"AgDK/gAA","Flags":19,"RawFlags":69699,"ParentIndex":0,"MasterIndex":0,"Namespace":null,"Alias":"","Statistics":{"RxPackets":0,"TxPackets":0,"RxBytes":0,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"ether","Protinfo":null,"OperState":0,"MulticastSnooping":false,"HelloTime":200},{"Index":85,"MTU":1500,"TxQLen":0,"Name":"eth0","HardwareAddr":"AgDK/gAA","Flags":19,"RawFlags":69699,"ParentIndex":86,"MasterIndex":3,"Namespace":null,"Alias":"","Statistics":{"RxPackets":2,"TxPackets":0,"RxBytes":200,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"ether","Protinfo":null,"OperState":6,"PeerName":""}]
2017-09-08 14:34:05.087916647 -0500 CDT:6554:cc-runtime:info:return bridgeNetworkPair
2017-09-08 14:34:05.092252339 -0500 CDT:6554:cc-runtime:info:launching qemu with: [-name pod-5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c -uuid 35626531-3235-3763-3333-396637333732 -machine pc,accel=kvm,kernel_irqchip,nvdimm -cpu host -qmp unix:/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/monitor.sock,server,nowait -qmp unix:/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/ctrl.sock,server,nowait -m 2048M,slots=2,maxmem=16947M -smp 8,cores=8,threads=1,sockets=1 -device virtio-9p-pci,fsdev=ctr-9p-0,mount_tag=ctr-rootfs-0 -fsdev local,id=ctr-9p-0,path=/var/lib/docker/overlay2/8dc7030d70f02c3a14b700b39b6eca6e59c607324154ba29458908ae6480a681/merged,security_model=none -device virtio-serial-pci,id=serial0 -device virtconsole,chardev=charconsole0,id=console0 -chardev socket,id=charconsole0,path=/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/console.sock,server,nowait -device nvdimm,id=nv0,memdev=mem0 -object memory-backend-file,id=mem0,mem-path=/usr/share/clear-containers/clear-containers.img,size=235929600 -device virtserialport,chardev=charch0,id=channel0,name=sh.hyper.channel.0 -chardev socket,id=charch0,path=/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/hyper.sock,server,nowait -device virtserialport,chardev=charch1,id=channel1,name=sh.hyper.channel.1 -chardev socket,id=charch1,path=/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/tty.sock,server,nowait -device virtio-9p-pci,fsdev=extra-9p-hyperShared,mount_tag=hyperShared -fsdev local,id=extra-9p-hyperShared,path=/tmp/hyper/shared/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c,security_model=none -device driver=virtio-net-pci,netdev=network-0,mac=02:42:ac:11:00:02 -netdev tap,id=network-0,ifname=tap0,downscript=no,script=no,vhost=on -rtc base=utc,driftfix=slew -global kvm-pit.lost_tick_policy=discard -vga none -no-user-config -nodefaults -nographic -daemonize -kernel /usr/share/clear-containers/vmlinuz.container -append root=/dev/pmem0p1 rootflags=dax,data=ordered,errors=remount-ro rw rootfstype=ext4 tsc=reliable no_timer_check rcupdate.rcu_expedited=1 i8042.direct=1 i8042.dumbkbd=1 i8042.nopnp=1 i8042.noaux=1 noreplace-smp reboot=k panic=1 console=hvc0 console=hvc1 initcall_debug iommu=off cryptomgr.notests net.ifnames=0 quiet systemd.show_status=false init=/usr/lib/systemd/systemd systemd.unit=clear-containers.target systemd.mask=systemd-networkd.service systemd.mask=systemd-networkd.socket ip=::::::5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c::off::]
cc-runtime:info:return bridgeNetworkPair
```

```
cc-runtime:info:cc-runtime (version 3.0.0-beta.2, commit 2259ee5a15b43a957feeb0c8b4e5b08746a14428-dirty) called as: [create --bundle /var/run/docker/libcontainerd/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c --console-socket /tmp/pty352205411/pty.sock --pid-file /run/docker/libcontainerd/containerd/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/init/pid 5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c]
cc-runtime:info:Using configuration file "/etc/clear-containers/configuration.toml"
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:No sockets from configuration
cc-runtime:info:Device details for container 5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c: Major:0, Minor:61, MountPoint:/var/lib/docker/overlay2/8dc7030d70f02c3a14b700b39b6eca6e59c607324154ba29458908ae6480a681/merged

cc-runtime:info:Calling bridgeNetworkPair
"lo","HardwareAddr":null,"Flags":5,"RawFlags":65609,"ParentIndex":0,"MasterIndex":0,"Namespace":null,"Alias":"","Statistics":{"RxPackets":0,"TxPackets":0,"RxBytes":0,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"loopback","Protinfo":null,"OperState":0},{"Index":2,"MTU":1500,"TxQLen":1000,"Name":"tap0","HardwareAddr":"VppqtMYf","Flags":19,"RawFlags":4163,"ParentIndex":0,"MasterIndex":3,"Namespace":null,"Alias":"","Statistics":{"RxPackets":0,"TxPackets":0,"RxBytes":0,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"ether","Protinfo":null,"OperState":0,"LinkType":"tun"},{"Index":3,"MTU":1500,"TxQLen":0,"Name":"br0","HardwareAddr":"AgDK/gAA","Flags":19,"RawFlags":69699,"ParentIndex":0,"MasterIndex":0,"Namespace":null,"Alias":"","Statistics":{"RxPackets":0,"TxPackets":0,"RxBytes":0,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"ether","Protinfo":null,"OperState":0,"MulticastSnooping":false,"HelloTime":200},{"Index":85,"MTU":1500,"TxQLen":0,"Name":"eth0","HardwareAddr":"AgDK/gAA","Flags":19,"RawFlags":69699,"ParentIndex":86,"MasterIndex":3,"Namespace":null,"Alias":"","Statistics":{"RxPackets":2,"TxPackets":0,"RxBytes":200,"TxBytes":0,"RxErrors":0,"TxErrors":0,"RxDropped":0,"TxDropped":0,"Multicast":0,"Collisions":0,"RxLengthErrors":0,"RxOverErrors":0,"RxCrcErrors":0,"RxFrameErrors":0,"RxFifoErrors":0,"RxMissedErrors":0,"TxAbortedErrors":0,"TxCarrierErrors":0,"TxFifoErrors":0,"TxHeartbeatErrors":0,"TxWindowErrors":0,"RxCompressed":0,"TxCompressed":0},"Promisc":0,"Xdp":null,"EncapType":"ether","Protinfo":null,"OperState":6,"PeerName":""}]
cc-runtime:info:return bridgeNetworkPair
cc-runtime:info:launching qemu with: [-name pod-5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c -uuid 35626531-3235-3763-3333-396637333732 -machine pc,accel=kvm,kernel_irqchip,nvdimm -cpu host -qmp unix:/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/monitor.sock,server,nowait -qmp unix:/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/ctrl.sock,server,nowait -m 2048M,slots=2,maxmem=16947M -smp 8,cores=8,threads=1,sockets=1 -device virtio-9p-pci,fsdev=ctr-9p-0,mount_tag=ctr-rootfs-0 -fsdev local,id=ctr-9p-0,path=/var/lib/docker/overlay2/8dc7030d70f02c3a14b700b39b6eca6e59c607324154ba29458908ae6480a681/merged,security_model=none -device virtio-serial-pci,id=serial0 -device virtconsole,chardev=charconsole0,id=console0 -chardev socket,id=charconsole0,path=/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/console.sock,server,nowait -device nvdimm,id=nv0,memdev=mem0 -object memory-backend-file,id=mem0,mem-path=/usr/share/clear-containers/clear-containers.img,size=235929600 -device virtserialport,chardev=charch0,id=channel0,name=sh.hyper.channel.0 -chardev socket,id=charch0,path=/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/hyper.sock,server,nowait -device virtserialport,chardev=charch1,id=channel1,name=sh.hyper.channel.1 -chardev socket,id=charch1,path=/run/virtcontainers/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c/tty.sock,server,nowait -device virtio-9p-pci,fsdev=extra-9p-hyperShared,mount_tag=hyperShared -fsdev local,id=extra-9p-hyperShared,path=/tmp/hyper/shared/pods/5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c,security_model=none -device driver=virtio-net-pci,netdev=network-0,mac=02:42:ac:11:00:02 -netdev tap,id=network-0,ifname=tap0,downscript=no,script=no,vhost=on -rtc base=utc,driftfix=slew -global kvm-pit.lost_tick_policy=discard -vga none -no-user-config -nodefaults -nographic -daemonize -kernel /usr/share/clear-containers/vmlinuz.container -append root=/dev/pmem0p1 rootflags=dax,data=ordered,errors=remount-ro rw rootfstype=ext4 tsc=reliable no_timer_check rcupdate.rcu_expedited=1 i8042.direct=1 i8042.dumbkbd=1 i8042.nopnp=1 i8042.noaux=1 noreplace-smp reboot=k panic=1 console=hvc0 console=hvc1 initcall_debug iommu=off cryptomgr.notests net.ifnames=0 quiet systemd.show_status=false init=/usr/lib/systemd/systemd systemd.unit=clear-containers.target systemd.mask=systemd-networkd.service systemd.mask=systemd-networkd.socket ip=::::::5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c::off::]
cc-runtime:info:{"QMP": {"version": {"qemu": {"micro": 0, "minor": 7, "major": 2}, "package": ""}, "capabilities": []}}
cc-runtime:info:QMP version 2.7.0
cc-runtime:info:QMP capabilities []
cc-runtime:info:{"execute":"qmp_capabilities"}
cc-runtime:info:{"return": {}}
cc-runtime:info:VM started
cc-runtime:info:Shim(s) started
cc-runtime:info:cgroup path /sys/fs/cgroup/memory mounted
cc-runtime:info:cgroup path /sys/fs/cgroup/cpu mounted
cc-runtime:info:cgroup path /sys/fs/cgroup/pids mounted
cc-runtime:info:cgroup path /sys/fs/cgroup/blkio mounted
c-runtime:info:cc-runtime (version 3.0.0-beta.2, commit 2259ee5a15b43a957feeb0c8b4e5b08746a14428-dirty) called as: [start 5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c]
cc-runtime:info:Using configuration file "/etc/clear-containers/configuration.toml"
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Running inside a VM = %tfalse
c-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Stopping shim PID 6594
cc-runtime:error:ERROR received from Hyperstart
cc-runtime:info:cc-runtime (version 3.0.0-beta.2, commit 2259ee5a15b43a957feeb0c8b4e5b08746a14428-dirty) called as: [kill --all 5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c SIGKILL]
cc-runtime:info:cc-runtime (version 3.0.0-beta.2, commit 2259ee5a15b43a957feeb0c8b4e5b08746a14428-dirty) called as: [delete 5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c]
cc-runtime:info:Using configuration file "/etc/clear-containers/configuration.toml"
cc-runtime:info:Using configuration file "/etc/clear-containers/configuration.toml"
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Running inside a VM = %tfalse
c-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Running inside a VM = %tfalse
c-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Stopping shim PID 6594
cc-runtime:info:cc-runtime (version 3.0.0-beta.2, commit 2259ee5a15b43a957feeb0c8b4e5b08746a14428-dirty) called as: [delete 5be1257c339f737242f057163eadc0386fc4227710122b0eaaf300621166e56c]
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Using configuration file "/etc/clear-containers/configuration.toml"
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:Stopping shim PID 6594
cc-runtime:info:Shim(s) stopped
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:info:{"QMP": {"version": {"qemu": {"micro": 0, "minor": 7, "major": 2}, "package": ""}, "capabilities": []}}
cc-runtime:info:{"execute":"qmp_capabilities"}
cc-runtime:info:{"return": {}}
cc-runtime:info:{"execute":"quit"}
cc-runtime:info:{"return": {}}
cc-runtime:info:{"timestamp": {"seconds": 1504899246, "microseconds": 38858}, "event": "SHUTDOWN"}
cc-runtime:error:Could not get veth interface: LinkByName() failed for veth name eth0: Link not found
cc-runtime:info:Running inside a VM = %tfalse
cc-runtime:error:Can not move from stopped to stopped
```

**Comment by jcvenegas** [comment]
@devimc  @sameo the veth pair is being deleted as docker ( we dont own that veth pair but use it for our networking). Added a warning instead of fail.) 
Now this patch + locking the container to perform an operation allow to properly delete a container (https://github.com/clearcontainers/runtime/issues/541).
```
2017-09-11 11:05:08.482744767 -0500 CDT:15550:cc-runtime:info:cc-runtime (version 3.0.0-beta.2, commit 6ac3e8c2db51b53efe5c3d56f3e4455c4c6ba2e2-dirty) called as: [delete 78219321addc1a01734c5198e3899e2b55df772ea179832f7324e7a995f25302]
2017-09-11 11:05:08.493875052 -0500 CDT:15550:cc-runtime:info:Using configuration file "/etc/clear-containers/configuration.toml"
2017-09-11 11:05:08.51783552 -0500 CDT:15550:cc-runtime:info:Acquiring lock on /var/78219321addc1a01734c5198e3899e2b55df772ea179832f7324e7a995f25302
2017-09-11 11:05:08.554298336 -0500 CDT:15550:cc-runtime:info:Acquired filelock on /var/78219321addc1a01734c5198e3899e2b55df772ea179832f7324e7a995f25302
2017-09-11 11:05:08.66881772 -0500 CDT:15550:cc-runtime:info:Stopping shim PID 15516
2017-09-11 11:05:08.692670233 -0500 CDT:15550:cc-runtime:info:Shim(s) stopped
2017-09-11 11:05:08.7124548 -0500 CDT:15550:cc-runtime:info:{"QMP": {"version": {"qemu": {"micro": 0, "minor": 7, "major": 2}, "package": ""}, "capabilities": []}}
2017-09-11 11:05:08.736121947 -0500 CDT:15550:cc-runtime:info:{"execute":"qmp_capabilities"}
2017-09-11 11:05:08.75838617 -0500 CDT:15550:cc-runtime:info:{"return": {}}
2017-09-11 11:05:08.78132173 -0500 CDT:15550:cc-runtime:info:{"execute":"quit"}
2017-09-11 11:05:08.796234866 -0500 CDT:15550:cc-runtime:info:{"return": {}}
2017-09-11 11:05:08.812017491 -0500 CDT:15550:cc-runtime:info:{"timestamp": {"seconds": 1505145908, "microseconds": 796119}, "event": "SHUTDOWN"}
2017-09-11 11:05:08.892530011 -0500 CDT:15550:cc-runtime:warning:Could not get veth interface: %sLinkByName() failed for veth name eth0: Link not found
2017-09-11 11:05:08.934845059 -0500 CDT:15550:cc-runtime:info:cgroup path /sys/fs/cgroup/memory mounted
2017-09-11 11:05:08.944425245 -0500 CDT:15550:cc-runtime:info:cgroup path /sys/fs/cgroup/cpu mounted
2017-09-11 11:05:08.958504414 -0500 CDT:15550:cc-runtime:info:cgroup path /sys/fs/cgroup/pids mounted
2017-09-11 11:05:08.969400477 -0500 CDT:15550:cc-runtime:info:cgroup path /sys/fs/cgroup/blkio mounted
```

File: pkg/oci/utils.go [hunk 1]
```diff
@@ -33,8 +33,8 @@ var (
 	// ErrNoLinux is an error for missing Linux sections in the OCI configuration file.
 	ErrNoLinux = errors.New("missing Linux section")
 
-	// ConfigPathKey is the annotation key to fetch the OCI configuration file path.
-	ConfigPathKey = "com.github.containers.virtcontainers.pkg.oci.config_path"
+	// ConfigJSONKey is the annotation key to fetch the OCI configuration.
+	ConfigJSONKey = "com.github.containers.virtcontainers.pkg.oci.config"
```
Comment (sameo) [inline pkg/oci/utils.go]: So by saving the entire JSON, we protect ourselves against the case where the failure happens before we could store the config.json on disk ?

File: hyperstart.go [hunk 2]
```diff
@@ -560,6 +576,8 @@ func (h *hyper) startOneContainer(pod Pod, c Container) error {
 		return err
 	}
 
+	//Container Started, not need to stop shim
```
Comment (sameo) [inline hyperstart.go]: Nit:
```
// Container started, no need to stop the shim
```

File: hyperstart.go [hunk 3]
```diff
@@ -510,6 +509,23 @@ func (h *hyper) startPauseContainer(podID string) error {
...
+	workloadStarted := false
+
+	//In case cannot start the container, stop the shim associated with it
+	defer func() {
+		//NOTE: workloadStarted must be false in case shim being stopped
+		if workloadStarted == true {
+			return
+		}
+
+		virtLog.Debug("failed start container, stoping shim")
```
Comment (sameo) [inline hyperstart.go]: This should be a warning.

File: pkg/oci/utils.go (hunk 1 above)
Comment (jcvenegas) [reply pkg/oci/utils.go]: We protect ourselves by not relaying in the config.json from the bundle in oci. The runtime would need to save all it needs in its state file. In this case the state is the vircontainers state file and the runtime perform some operations using the information provides in this json.

File: hyperstart.go [hunk 4]
```diff
@@ -510,6 +509,23 @@ func (h *hyper) startPauseContainer(podID string) error {
...
+		if err := stopShim(c.process.Pid); err != nil {
```
Comment (sameo) [inline hyperstart.go]: So the logic makes perfect sense and it's definitely a case we're not handling properly.
However, I would like it to be handled at a higher level, i.e. not at the hyperstart one, because this logic applies to all agent: When we fail to start a container, we should remove its shim.

So what I'm saying is that the shim deletion/stopping should be done from the `pod.go` and `container.go` files. We almost do that from `container.go`:

```Go
	err = c.pod.agent.startContainer(*(c.pod), *c)
	if err != nil {
		c.stop()
		return err
	}
```

But we don't for `pod.go`. `pod.stop()` should be called and it should take care of cleaning any dangling shim.

**Comment by sameo** [comment]
Patches 1 and 3 make sense to me, I would just like to see the second one reworked to not be hyperstart specific.
I'd like to get this merged before we tag virtcontainers 1.0.0.

File: container.go [hunk 5]
```diff
@@ -436,17 +436,11 @@ func (c *Container) fetchState(cmd string) (State, error) {
...
 		return State{}, fmt.Errorf("Cmd cannot be empty")
 	}
-
-	state, err := c.pod.storage.fetchPodState(c.pod.id)
-	if err != nil {
-		return State{}, err
-	}
-
-	if state.State != StateRunning {
+	if c.pod.state.State != StateRunning {
```
Comment (sameo) [inline container.go]: Could you explain why we're no longer fetching the state from storage ?

File: container.go [hunk 6]
```diff
@@ -488,9 +482,13 @@ func (c *Container) start() error {
 		}
 	}
 
-	err = c.pod.agent.startContainer(*(c.pod), *c)
-	if err != nil {
-		c.stop()
+	if err = c.pod.agent.startContainer(*(c.pod), *c); err != nil {
+
```
Comment (sameo) [inline container.go]: nitpick: Unneeded new line.

File: pod.go [hunk 7]
```diff
@@ -795,14 +790,17 @@ func (p *Pod) start() error {
...
 	if err := p.startSetStates(); err != nil {
 		return err
 	}
 
+	for _, c := range p.containers {
+		if err := c.start(); err != nil {
+			return err
+		}
+	}
+
```
Comment (sameo) [inline pod.go]: So you're changing the container starting and state setting logic here, by moving it outside of hyperstart into pod.go. I like it, but it should be part of a separate patch on this PR.

File: container.go (hunk 5 above)
Comment (jcvenegas) [reply container.go]: before this line being executed I changed, the pod state by using `p.startSetStates()` which changes the state only in memory but in in storage, I think we should relay in the current state of the pod object that is in memory. Otherwise I could do `p.startSetStates()`  + a method to save the pod state in storage ( Taking advantage of this comment, what is that method ? )

File: Makefile [hunk 8]
```diff
@@ -46,7 +46,7 @@ binaries: virtc pause hook shim
 # Tests
 #
 
-check: check-go-static check-go-test
+check: check-go-test
```
[the remark this answers is not in the document]
Comment (jcvenegas) [reply Makefile]: Ups, reverting.

File: pod.go (hunk 7 above)
Comment (jcvenegas) [reply pod.go]: sure, @sameo at some point, I dont know if this is the correct time but I think we need to change the pod.start method to only start the pod ( through the agent ) and start the `pause` container, the rest of containers must be started by `StartContaine`

File: container.go (hunk 5 above)
Comment (sameo) [reply container.go]: > before this line being executed I changed, the pod state by using `p.startSetStates()`

Where do you do that ?

> `p.startSetStates()` + a method to save the pod state in storage ( Taking advantage of this comment, what is that method ? )

`p.startSetStates() ` already stores the pod state in storage.

File: pod.go (hunk 7 above)
Comment (sameo) [reply pod.go]: That's almost what I'm asking you to do in a separate patch, except for separating the pause container from the other containers.

Any reason why we'd need to start the pause one at pod starting time ?

Comment (jcvenegas) [reply pod.go]: @sameo I spitted the commit. I hope I helps for readability .
`Any reason why we'd need to start the pause one at pod starting time ?`
 it is not something I want but we already do, https://github.com/containers/virtcontainers/blob/master/hyperstart.go#L453, notice that this PR does not move that line to pod level. I dont see anything in the agent interface to to something like `h.stopPauseContainer`, so I am guessing that is hyperstart specific ? .

File: container.go (hunk 5 above)
Comment (sameo) [inline container.go]: @jcvenega Sorry but I still don't understand why we need to fetch a container state from a pod that's not running yet. This should not happen, and if it does, the container state should be set to `ready` at most.

In the commit log, you need to explain why we need this, on not only how you do it.

File: pod.go [hunk 9]
```diff
@@ -699,11 +699,6 @@ func (p *Pod) startSetStates() error {
 		return err
 	}
 
-	err = p.setContainersState(podState.State)
```
Comment (sameo) [inline pod.go]: If we removing the container state setting, the function needs to be renamed to `startSetState()`.

File: container.go (hunk 5 above)
Comment (jcvenegas) [reply container.go]: >Where do you do that ?

https://github.com/containers/virtcontainers/pull/360/files#diff-c7bc83b548b0dcb1a7e055bb0d07779cR793
```Go
	// Pod is started
	if err := p.startSetStates(); err != nil {
		return err
	}

	for _, c := range p.containers {
		if err := c.start(); err != nil {
			return err
		}
	}
```
1. Start pod.
2. Change the state of pod to started calling `p.startSetStates`
3. For each container in pod : do `cotainer.start()` 
`containter.start()` will call `container.fetchState()` which require the pod state being in `running` state. But I got an error saying pod was not running. 

Let me check again and see if it is working correctly by getting from `storage`.

Comment (jcvenegas) [reply container.go]: @sameo  you was right, I is not needed, my bad,  commit removed.

File: pod.go (hunk 9 above)
Comment (jcvenegas) [reply pod.go]: renamed

**Comment by jcvenegas** [comment]
I tested this PR with cc-runtime unit tests and functional, crio, docker. Only runtime unit-test need to be updated to work with this new changes.
Here the needed changes.
https://github.com/jcvenegas/runtime/commit/ca2b7f95d5462f6b0cc84306fdb8638d533f353f

File: container.go [hunk 10]
```diff
@@ -528,6 +531,18 @@ func (c *Container) stop() error {
 		return err
 	}
 
+	defer func() {
```
Comment (sboeuf) [inline container.go]: I don't think this part was needed since the shim is started by `createContainer()` and stopped by `delete()`. It was on purpose that we were stopping the shim in the `delete()`, in order to mirror the create behavior.
I don't think this code part hurt but it kind of duplicates the stop of the shim, and it actually slow down the stop action.
